### PR TITLE
Feat: better (eval) error messages/context

### LINKED ISSUE #783: Error context is lost
This is Alpine v2.7.0.

If I do something like this:

```
 <template x-for="item in thisdoesnotexist" :key="item">
<div></div>
</template>
```

I get this not very useful error in the console:

```html
[Error] TypeError: undefined is not an object (evaluating 'items.forEach')
	(anonymous function) (alpine.js:1762)
```

In non-trivial setups, guessing where the faulty template code is, is extremely hard.

I remember in a not-so-distant Alpine version that I saw some more context (as in: the actual script string that failed to evaluate).

### PR DESCRIPTION
**Pull request by HugoDF**
Closes #783 

from these discussion points on the other PR:
- https://github.com/alpinejs/alpine/pull/862#issuecomment-732275009 
- https://github.com/alpinejs/alpine/pull/862#issuecomment-732293276

Implement the suggestion from the latter comment.

Errors are logged as follows:

Screenshot 2020-11-23 at 18 14 42

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -65,6 +65,24 @@ export function debounce(func, wait) {
     }
 }
 
+const handleError = (el, expression, error) => {
+    console.error(`Alpine: error in expression "${expression}" in component:`, el, `due to "${error}"`);
+    if (!isTesting()) {
```
Comment (HugoDF) [inline src/utils.js:70]: Did this so that the app doesn't throw during tests eg. when testing event handlers

File: src/utils.js [hunk 2]
```diff
@@ -65,6 +65,24 @@ export function debounce(func, wait) {
     }
 }
 
+const handleError = (el, expression, error) => {
+    console.error(`Alpine: error in expression "${expression}" in component:`, el, `due to "${error}"`);
+    if (! isTesting()) {
+        throw error;
+    }
```
Comment (HugoDF) [inline src/utils.js:72]: Avoid throwing during tests eg. when testing event handlers if we throw it's basically untestable

File: src/utils.js [hunk 3]
```diff
@@ -65,6 +65,24 @@ export function debounce(func, wait) {
...
+}
+
+export function tryCatch(cb, { el, expression }) {
+    try {
+        const value = cb();
+        return value instanceof Promise
+            ? value.catch((e) => handleError(el, expression, e))
+            : value;
+    } catch (e) {
+        handleError(el, expression, e)
```
Comment (HugoDF) [inline src/utils.js:82]: handles sync errors

File: src/utils.js [hunk 4]
```diff
@@ -65,6 +65,24 @@ export function debounce(func, wait) {
...
+    if (! isTesting()) {
+        throw error;
+    }
+}
+
+export function tryCatch(cb, { el, expression }) {
+    try {
+        const value = cb();
+        return value instanceof Promise
+            ? value.catch((e) => handleError(el, expression, e))
```
Comment (HugoDF) [inline src/utils.js:79]: handle async errors eg. `saferEvalNoReturn` when it's an AsyncFunction

**Comment by calebporzio** [comment]
Thanks @HugoDF, this is looking good. I like how you are also reporting the expression that caused the breakage.

One thought. In the case of x-spread (I think), sometimes the `expression` passed to `saferEval` is a function, and not a string. In that case, the concatenation will likely fail. Can you check on that?

Sorry for putting you through the ringer on this.

Thanks again!

**Comment by HugoDF** [comment]
> Thanks @HugoDF, this is looking good. I like how you are also reporting the expression that caused the breakage.
> 
> One thought. In the case of x-spread (I think), sometimes the `expression` passed to `saferEval` is a function, and not a string. In that case, the concatenation will likely fail. Can you check on that?
> 
> Sorry for putting you through the ringer on this.
> 
> Thanks again!

I think it should work, let me add a unit test and example though 😄

**Comment by HugoDF** [comment]
🤔 we're going to be up to 4 `tryCatch`-es, at that point I reckon it might be more future-proof to pass `el` into `saferEval`/`saferEvalNoReturn` and do the error reporting there.

That's a pretty trivial refactor in the end, a case of removing `tryCatch` callsites and instead passing `el` into the eval functions.

**Comment by HugoDF** [comment]
@calebporzio I've moved the `tryCatch` into saferEval/saferEvalNoReturn see this PR https://github.com/HugoDF/alpine/pull/2/files (it's on my repo but will get reflected into this PR if you think it makes sense)

we save 0.1kb but more importantly we only have the `tryCatch` calls in 2 spots (downside is that we have to pass `el` to saferEval/saferEvalNoReturn)

**Comment by calebporzio** [comment]
No, I like the direction. Good call

**Comment by HugoDF** [comment]
> No, I like the direction. Good call

cool, should be ready for review then 👍

**Comment by calebporzio** [comment]
Thanks for this Hugo! Going to be extremely helpful for users

**Pull request merged**